### ISSUE TITLE
[BUG]: Problem with "free(): invalid pointer" after returning 2d numpy

### ISSUE THREAD

**Issue by gregogiudici** [wjakob/nanobind#716]
### Problem description

I created a `Stretch `class with nanobind. The class has a `process `method that receives a numpy array and returns a new numpy array. 
When I test it with the following code, everything works:
```
import numpy as np
import pystretch

if __name__ == "__main__":
      x = np.random.rand(1,44100)
      ps = pystretch.SignalsmithStretch.Stretch()
      ps.preset(1,44100)
      ps.timeFactor = 2
      y = ps.process(x) 
      print(y.shape) # (1, 88200) - No errors
```
The problem is when I try to run the `process `method multiple times.
```
ps = pystretch.SignalsmithStretch.Stretch()
ps.preset(1,44100)
ps.timeFactor = 2
x1 = np.random.rand(1,44100)
x2 = np.random.rand(1,22050)
y1 = ps.process(x1)
y2 = ps.process(x2)
print(y1.shape)
print(y2.shape)
```
I receive the following error:
```
(1, 88200)
(1, 44100)
free(): invalid pointer 
Aborted
```

### Reproducible example code

```text
struct Stretch {
private:
    signalsmith::stretch::SignalsmithStretch<float> stretch_;
public:
    Stretch() : stretch_() {}

    // === Attributes
    float sampleRate_;
    float timeFactor_ = 1.f;

   // ..... other stuff ....

    // === Processing ===
    nb::ndarray<nb::numpy, float, nb::ndim<2>> process(nb::ndarray<nb::numpy, float, nb::ndim<2>> audio_input) {
        auto inData = audio_input.data();

        if (audio_input.shape(0) > 2) {
            throw std::runtime_error("Only mono or stereo audio is supported. The input should have shape (1,samples) or (2,samples).");
        }

        size_t numChannels = audio_input.shape(0);
        size_t inputSamples = audio_input.shape(1);
        
        // Padding for latency
        size_t paddedInputSamples  = inputSamples + stretch_.inputLatency();
        int tailSamples = stretch_.outputLatency();
        int outputSamples = std::round(inputSamples * timeFactor_);
        size_t paddedOutputSamples = outputSamples + tailSamples;

        // Allocate and initialize buffers
        float** inputChannels = new float*[numChannels];
        float** outputChannels = new float*[numChannels];
        
        for (size_t i = 0; i < numChannels; ++i) {
            inputChannels[i] = new float[paddedInputSamples]();
            outputChannels[i] = new float[paddedOutputSamples]();
        }

        // Copy input audio to input_data
        for (size_t i = 0; i < numChannels; ++i) {
            std::copy(inData + i*inputSamples , inData + (i+1)*inputSamples , inputChannels[i]);
        }

        // Wrap input/output buffer with Buffer class (for offset reading/writing)
        Buffer<float> inBuffer(inputChannels, paddedInputSamples);
        Buffer<float> outBuffer(outputChannels, outputSamples);

        // Set offset of inBuffer
        inBuffer.setOffset(stretch_.inputLatency());

        // PROCESSING
        stretch_.process(inBuffer, inputSamples , outBuffer, outputSamples);

        // Read the last bit of output without providing any further input
        outBuffer.setOffset(outputSamples);
        stretch_.flush(outBuffer, tailSamples);
        outBuffer.setOffset(tailSamples);

        // Prepare output data
        size_t outShape[2] = {numChannels, outputSamples};
        float* outData = new float[numChannels * outShape[1]];

        // Copy output data to outData
        for (size_t i = 0; i < numChannels; ++i) {
            std::copy(outputChannels[i], outputChannels[i] + outputSamples, outData + i * outputSamples);
        }

        // Clean up
        for (size_t i = 0; i < numChannels; ++i) {
            delete[] inputChannels[i];
            delete[] outputChannels[i];
        }
        delete[] inputChannels;
        delete[] outputChannels;

        // Delete 'outData' when the 'owner' capsule expires
        nb::capsule owner(outData, [](void *p) noexcept {
            delete[] static_cast<float*>(p);
        });

        // Create the output ndarray
        return nb::ndarray<nb::numpy, float, nb::ndim<2>>(outData, 2, outShape, owner);
    }
};


NB_MODULE(SignalsmithStretch, m) {
    nb::class_<Stretch>(m, "Stretch")
        .def(nb::init<>())
        .def_rw("timeFactor", &Stretch<Sample>::timeFactor_)
        // ... Other methods ...
        // Processing   
        .def("process", &Stretch<Sample>::process,
            "audio_input"_a);
}
```

**Comment by wjakob** [comment]
Hi @gregogiudici,

this is not a reproducible example. Can you please make a compact reproducer that doesn't reference any external library code?

**Comment by gregogiudici** [comment]
Hi @wjakob,

Thank you for your quick reply! I discovered that I made a mistake. While creating a more compact example as you suggested, I discovered that the problem was related to the external library I was using, not to nanobind itself (I still need to practice with c++). After fixing the problem, the error no longer occurred.

Thank you again for your time and I also want to express my appreciation for nanobind. It is a fantastic library!
